protocol/message_sender.go

```diff
@@ -0,0 +1,250 @@
+
+	return nil
+}
+
+func (m *messageSenderImpl) messageSenderForPeer(ctx context.Context, p peer.ID) (*peerMessageSender, error) {
+	m.smlk.Lock()
+	ms, ok := m.strmap[p]
+	if ok {
+		m.smlk.Unlock()
+		return ms, nil
+	}
+	ms = &peerMessageSender{p: p, m: m, lk: newCtxMutex()}
+	m.strmap[p] = ms
+	m.smlk.Unlock()
+
+	if err := ms.prepOrInvalidate(ctx); err != nil {
+		m.smlk.Lock()
+		defer m.smlk.Unlock()
+
+		if msCur, ok := m.strmap[p]; ok {
+			// Changed. Use the new one, old one is invalid and
+			// not in the map so we can just throw it away.
+			if ms != msCur {
+				return msCur, nil
+			}
+			// Not changed, remove the now invalid stream from the
+			// map.
+			delete(m.strmap, p)
+		}
+		// Invalid but not in map. Must have been removed by a disconnect.
+		return nil, err
+	}
+	// All ready to go.
+	return ms, nil
+}
+
+// peerMessageSender is responsible for sending requests and messages to a particular peer
+type peerMessageSender struct {
+	s  network.Stream
+	r  msgio.ReadCloser
+	lk ctxMutex
+	p  peer.ID
+	m  *messageSenderImpl
+
+	invalid   bool
+	singleMes int
+}
+
+// invalidate is called before this peerMessageSender is removed from the strmap.
+// It prevents the peerMessageSender from being reused/reinitialized and then
+// forgotten (leaving the stream open).
+func (ms *peerMessageSender) invalidate() {
+	ms.invalid = true
+	if ms.s != nil {
+		_ = ms.s.Reset()
+		ms.s = nil
+	}
+}
+
+func (ms *peerMessageSender) prepOrInvalidate(ctx context.Context) error {
+	if err := ms.lk.Lock(ctx); err != nil {
+		return err
+	}
+	defer ms.lk.Unlock()
+
+	if err := ms.prep(ctx); err != nil {
+		ms.invalidate()
+		return err
+	}
+	return nil
+}
+
+func (ms *peerMessageSender) prep(ctx context.Context) error {
+	if ms.invalid {
+		return fmt.Errorf("message sender has been invalidated")
+	}
+	if ms.s != nil {
+		return nil
+	}
+
+	// We only want to speak to peers using our primary protocols. We do not want to query any peer that only speaks
+	// one of the secondary "server" protocols that we happen to support (e.g. older nodes that we can respond to for
+	// backwards compatibility reasons).
+	nstr, err := ms.m.host.NewStream(ctx, ms.p, ms.m.protocols...)
+	if err != nil {
+		return err
+	}
+
+	ms.r = msgio.NewVarintReaderSize(nstr, network.MessageSizeMax)
+	ms.s = nstr
+
+	return nil
+}
+
+func (ms *peerMessageSender) SendMessage(ctx context.Context, pmes *pb.Message) error {
+	if err := ms.lk.Lock(ctx); err != nil {
+		return err
+	}
+	defer ms.lk.Unlock()
+
+	if err := ms.prep(ctx); err != nil {
+		return err
+	}
+
+	if err := ms.writeMsg(pmes); err != nil {
+		_ = ms.s.Reset()
+		ms.s = nil
+
+		log.Debugw("error writing message", "error", err, "retrying", true)
+		return err
+	}
+
+	return nil
+}
+
+func (ms *peerMessageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb.Message, error) {
+	if err := ms.lk.Lock(ctx); err != nil {
+		return nil, err
+	}
+	defer ms.lk.Unlock()
+
+	if err := ms.prep(ctx); err != nil {
```

Comment:
SendMessage and SendRequest overlap in the first half of their implementation. Worth folding this half it into a separate function?